feat(cloudTags): add custom colors support for cloudTags on card_tags widget & tags page Pr_Satus(Finished)

Void4m0n opened the pull request:
# PR Motivation
In the current implementation, both the tags page and the `card_tags` widget (when `color` is enabled) render their tag lists using the `cloudTags` helper, which only allows generating tags with random colors. This can cause visual inconsistencies when users style their site around a specific color palette, breaking overall design consistency.

# PR Summary
This PR adds an optional custom palette mode to the `cloudTags` helper. When `colormode` is set to `custom` and `custom_colors` are provided, `cloudTags` doesnt hardcodes colors inline; instead it assigns cyclic `tag-color-*` classes (repeating once it reaches the end of the palette) and leaves the actual color values to CSS. The palette is read from the theme config and generated in Stylus, which makes it easy to target from CSS. If `colormode` is not set (or isn’t `custom`), the helper remains fully backwards compatible and continues using the existing `getRandomColor()` inline-color behavior. The tags page is also aligned with the same theme configuration used by the `card_tags` widget so both outputs stay consistent.

## Example
## Config
```yml
aside:
  card_tags:
    enable: true
    # If set 0 will show all
    limit: 40
    color: true
    colormode: custom
    custom_colors:
      - "rgb(111, 79, 140)"
      - "rgba(94, 62, 126, 0.9)"
```
## Helper

- Output example:

```html
<a href="/tags/Tag-0" class="tag-cloud-item tag-color-0" style="font-size: 1.5em;">TAG 0</a>
<a href="/tags/Tag-1" class="tag-cloud-item tag-color-1" style="font-size: 1.5em;">TAG 1</a>
```
- Cycles classes using idx % custom_colors.length.
- No inline background-color / color is generated in this mode.
- Default mode (unchanged): still uses getRandomColor() and inline color styles as before.

## Templates

tags.pug now passes colormode and custom_colors from theme config (same source as card_tags).

## Stylus / CSS

- Reads palette from theme config: hexo-config('aside.card_tags.custom_colors')
- Generates .tag-color-* slot classes and maps them to --tag-cloud-color.
- Applies the variable depending on the container:
  - .tag-cloud-list .tag-cloud-item → background-color: var(--tag-cloud-color)
  - .card-tag-cloud .tag-cloud-item → color: var(--tag-cloud-color)

Example output CSS (for illustration):
```css
.tag-cloud-list .tag-cloud-item {
  background-color: var(--tag-cloud-color);
}
.card-tag-cloud .tag-cloud-item {
  color: var(--tag-cloud-color);
}
.tag-cloud-item.tag-color-0 {
  --tag-cloud-color: #6f4f8c;
}
.tag-cloud-item.tag-color-1 {
  --tag-cloud-color: rgba(94,62,126,0.9);
}
```
# Documentation Suggestion
## card_tags Widget (Aside)
[Current Docs](https://butterfly.js.org/en/posts/butterfly-docs-en-theme-config/?highlight=card_tag#Aside)

I suggest updating the `card_tags` section by adding `card_tags.colormode` and `card_tags.custom_colors`:

Syntax | Explanation
-- | --
card_tags.enable | Whether to display the tags card
card_tags.limit | Number of tags to display, 0 for all
card_tags.color | Whether to display tag colors.
card_tags.orderby | Tag sorting method: random / name / length
card_tags.order | Sorting method, 1 for ascending, -1 for descending
card_tags.colormode | [Optional] Color mode for tags when `card_tags.color: true`. Set to `custom` to use user-defined colors. If not set (or set to random), random colors are used.
card_tags.custom_colors | [Optional] Custom colors used when `card_tags.colormode: custom`. Accepts a single color string (applies the same color to all tags) or a YAML sequence (block sequence using `-` or flow sequence using `[...]`) to cycle colors repeatedly. Note: Hex colors must be quoted in YAML (since `#` starts a comment).

# Visual Examples
## Current Workflow
### Aside card_tags
#### Color False
<img width="671" height="208" alt="aside_tags_yaml_old_workflow" src="https://github.com/user-attachments/assets/16503f72-36e7-4066-be21-b1918b53c45c" />
<img width="1692" height="1053" alt="render_aside_tags_yaml_old_workflow" src="https://github.com/user-attachments/assets/a1355bfb-6b1b-4978-be08-a8c2305ed9e5" />

#### Color True
<img width="695" height="208" alt="aside_tags_color_true_yaml_old_workflow" src="https://github.com/user-attachments/assets/cf9b1be2-72f7-4f0e-85c3-22a49e834536" />
<img width="1597" height="1109" alt="render_aside_tags_color_old_workflow" src="https://github.com/user-attachments/assets/67becf24-74d9-4683-9770-25ecc0be9800" />
<img width="1447" height="685" alt="default_workflow" src="https://github.com/user-attachments/assets/09ae4568-3fcf-4a9e-8422-638ac39de781" />


## With PR Changes
### Aside card_tags
#### Colormode Custom
<img width="677" height="377" alt="aside_tags_color_custom_yaml" src="https://github.com/user-attachments/assets/3de8c69a-9b98-47d6-9b99-d5220f527f33" />
<img width="1236" height="836" alt="render_aside_tags_color_custom" src="https://github.com/user-attachments/assets/3918e8c6-9a40-4efb-a076-520e8009c5aa" />
<img width="1449" height="685" alt="render_array_custom_colors" src="https://github.com/user-attachments/assets/eea87dc6-4ad3-4fd6-9c26-7095f98829e9" />

#### Colormode random
<img width="680" height="377" alt="aside_tags_colormode_random_yaml" src="https://github.com/user-attachments/assets/65073fa8-9c27-408e-95b8-a310e351f54c" />
<img width="1225" height="834" alt="render_aside_tags_colormode_random" src="https://github.com/user-attachments/assets/ed711d40-53ff-4ac7-b0fb-ef557aec1990" />

#### Color False (Preserves the existing workflow.)
<img width="697" height="377" alt="aside_tags_color_false_yaml" src="https://github.com/user-attachments/assets/990f3817-497f-4405-8c79-39f066809898" />
<img width="1287" height="817" alt="render_aside_tags_color_false" src="https://github.com/user-attachments/assets/96251998-e64b-4de0-92d4-7f87e5fa68d6" />



P.S.: I'm not Chinese, so it's quite difficult for me to search through the discussions and PRs to see if something like this has already been talked about. I'm also not a developer, so my knowledge of these technologies is fairly limited, but I believe this change is useful. Thank you for this awesome theme.

Void4m0n committed:
feat(cloudTags): add custom colors support for cloudTags on tags page

Void4m0n committed:
feat(card_tags): add colormode/custom_colors to cloudTags call

Void4m0n committed:
refactor(tagcloud): switch custom_colors from inline styles to CSS classes

Void4m0n commented:
Hi @jerryc127, I wish you a happy New Year. 

This PR is ready for review.

jerryc127 commented:
感謝你的 PR！以下是我的一些建議，希望能讓這個功能更乾淨、彈性且維護成本更低：

1. **card_tags 的配置不應自動影響 tags 頁面**  
   目前 tags 頁面會直接套用 aside.card_tags 裡的 custom_colors，容易造成耦合。  
   建議讓 tags 頁面獨立控制，例如在 tags 頁的 front-matter 加上：
   ```yaml
   ---
   title: 標籤
   custom_colors:
     - "#ff6b6b"
     - "#4ecdc4"
     - "#45b7d1"
   ---
   ```
   這樣使用者可以為不同頁面設定不同的配色方案，更符合 Hexo 的頁面級配置習慣。

2. **可以移除 colormode 欄位**  
   當 `custom_colors` 有值時（無論是陣列還是單一字串），就直接視為使用自訂顏色；沒有值時才用隨機顏色。  
   這樣配置更簡潔，也不需要額外判斷 `colormode`，向後相容性更好。

3. **建議改用 inline style 而不是生成大量 class**  
   目前做法會動態產生很多 `.tag-color-0 ~ .tag-color-N` 的 CSS 規則，如果使用者配置很多顏色會讓 CSS 變很大，而且全站都會載入這些規則（即使某些頁面根本不用）。  
   推薦直接在 HTML 標籤上寫 `style="background-color: xxx;"`（或 `color: xxx;`），例如：
   ```html
   <a href="..." style="background-color: #ff6b6b;">標籤</a>
   ```
   優點：零額外 CSS、不影響其他頁面、顏色循環簡單（index % colors.length）、效能更好。

4. **請在 default_config.js 補上相關預設配置**  
   建議在 `scripts/common/default_config.js` 的 `aside.card_tags` 區塊加上預設配置：
   ```js
   card_tags: {
     enable: true,
     limit: 40,
     color: true,
     custom_colors: null, 
   },
   ```
---

Thank you for your PR! Here are some suggestions to make the feature cleaner, more flexible, and easier to maintain:

1. **card_tags config should not automatically affect the tags page**  
   Currently, the tags page directly inherits custom_colors from aside.card_tags, which creates coupling.  
   Suggestion: Make the tags page independent, e.g., allow configuration via front-matter:
   ```yaml
   ---
   title: Tags
   custom_colors:
     - "#ff6b6b"
     - "#4ecdc4"
     - "#45b7d1"
   ---
   ```
   This gives users more flexibility to use different color schemes on different pages, which aligns better with Hexo’s page-level config convention.

2. **colormode field can be removed**  
   If `custom_colors` has a value (array or single string), treat it as custom mode; otherwise fall back to random.  
   This makes the config simpler, eliminates the need for extra `colormode` checks, and improves backward compatibility.

3. **Recommend using inline style instead of generating many CSS classes**  
   The current approach generates lots of `.tag-color-0 ~ .tag-color-N` rules, which can bloat the CSS file if many colors are configured, and loads unnecessary CSS on pages that don’t even use tag clouds.  
   Suggestion: Directly apply `style="background-color: xxx;"` (or `color: xxx;`) on each tag element, e.g.:
   ```html
   <a href="..." style="background-color: #ff6b6b;">Tag</a>
   ```
   Advantages: No extra CSS, no impact on other pages, simple cycling (index % colors.length), and better performance.

4. **Please add default config examples in default_config.js**  
   Suggest adding default config values in `scripts/common/default_config.js` under `aside.card_tags` for easier discovery:
   ```js
   card_tags: {
     enable: true,
     limit: 40,
     color: true,
     custom_colors: null
   },
   ```

Void4m0n committed:
fix(tagcloud): independent tags page and card_tags, drop colormode, use inline css, keep tag-color classes

Void4m0n commented:
Thanks for the detailed feedback, I agree with your suggestions and I’ve applied them, with a small exception in point (3).

1) **Make tags page and card_tags Independent**  
The tags page no longer inherits colors from `aside.card_tags`. It now reads `custom_colors` from the Tags page front-matter, so users can define a page-specific palette independently from the sidebar widget.

2) **Remove `colormode`**  
I removed the `colormode` field. If `custom_colors` is provided (single string or array), `cloudTags` uses it as the custom palette; otherwise it falls back to the existing random color behavior.

3) **Inline styles vs CSS class generation**  
I agree with the inline-style approach and initially went with it. The main issue I ran into is that hardcoding colors inline makes it difficult to provide different palettes for light/dark themes, which can impact contrast and accessibility (e.g., Core Web Vitals / Lighthouse contrast checks).  
To keep the implementation lightweight while still giving users full control, I’m using inline colors as you suggested **and** keeping a cyclic `tag-color-*` class on each tag, but without generating any CSS rules. This allows end users to override colors per theme via their own CSS (using `!important` when needed to override inline styles).

4) **Default config example**  
I added `custom_colors: null` under `aside.card_tags` in `scripts/common/default_config.js` for easier discovery.
